### PR TITLE
Suggest ..= when someone tries to create an overflowing range

### LINKED ISSUE #109529: Suggest `..=` when someone tries to create an overflowing range
I tried this code:
```rust
for i in 0..256 as u8 {
        println!("{}", i);
}
```

I expected to see this happen: 
printing all numbers from 0 to 255 (inclusive)

Instead, this happened: 
```
error: literal out of range for `u8`
  --> src/main.rs:10:17
   |
10 |     for i in 0..256 as u8 {
   |                 ^^^
   |
   = note: the literal `256` does not fit into the type `u8` whose range is `0..=255`
   = note: `#[deny(overflowing_literals)]` on by default
```

`rustc --version --verbose`:
```
rustc 1.68.0 (2c8cc3432 2023-03-06)
binary: rustc
commit-hash: 2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74
commit-date: 2023-03-06
host: x86_64-unknown-linux-gnu
release: 1.68.0
LLVM version: 15.0.6
```

### PR DESCRIPTION
**Pull request by mu001999**
Fixes #109529

### CODE REVIEW

File: compiler/rustc_lint/src/types.rs [hunk 1]
```diff
@@ -136,6 +136,13 @@ fn lint_overflowing_range_endpoint<'tcx>(
     expr: &'tcx hir::Expr<'tcx>,
     ty: &str,
 ) -> bool {
+    let (expr, cast_ty) = if let Node::Expr(par_expr) = cx.tcx.hir().get(cx.tcx.hir().parent_id(expr.hir_id))
```
Comment (WaffleLapkin) [inline compiler/rustc_lint/src/types.rs:139]: Can you add a comment here describing what this if does? Something like
```rust
// Look past casts to support cases like `0..256 as u8`
```

File: compiler/rustc_lint/src/types.rs [hunk 2]
```diff
@@ -157,13 +164,19 @@ fn lint_overflowing_range_endpoint<'tcx>(
...
 
-    use rustc_ast::{LitIntType, LitKind};
-    let suffix = match lit.node {
-        LitKind::Int(_, LitIntType::Signed(s)) => s.name_str(),
-        LitKind::Int(_, LitIntType::Unsigned(s)) => s.name_str(),
-        LitKind::Int(_, LitIntType::Unsuffixed) => "",
-        _ => bug!(),
+    let suffix = if let Some(cast_ty) = cast_ty {
+        let Ok(ty) = cx.sess().source_map().span_to_snippet(cast_ty.span) else { return false };
+        format!(" as {}", ty)
```
Comment (WaffleLapkin) [inline compiler/rustc_lint/src/types.rs:169]: Why do we need to change `as` at all? can't we instead change `suggestion: struct_expr.span,` to something like `struct_expr.span.shrink_to_lo().to(eps[1].expr.span)` (i.e. suggest to change the `0..256` part, but not the `as ...`)?

File: compiler/rustc_lint/src/types.rs (hunk 1 above)
Comment (MU001999) [reply compiler/rustc_lint/src/types.rs:139]: Of course!

File: compiler/rustc_lint/src/types.rs (hunk 2 above)
Comment (MU001999) [reply compiler/rustc_lint/src/types.rs:169]: Got it! Thanks!

**Comment by leonardo-m** [comment]
This suggestion seems OK, but keep in mind that using ..= has performance consequences (it often/sometimes gives slower code).

File: compiler/rustc_lint/src/types.rs (hunk 2 above)
Comment (MU001999) [reply compiler/rustc_lint/src/types.rs]: I tried, but the suggestions don't work well if parentheses appear (suggestions will break parentheses).
In the latest commit, I separated the suggestions for the `=` and the literal.

File: compiler/rustc_lint/src/lints.rs [hunk 3]
```diff
@@ -1210,9 +1210,10 @@ impl<'a> DecorateLint<'a, ()> for DropGlue<'_> {
 #[diag(lint_range_endpoint_out_of_range)]
 pub struct RangeEndpointOutOfRange<'a> {
     pub ty: &'a str,
-    #[suggestion(code = "{start}..={literal}{suffix}", applicability = "machine-applicable")]
-    pub suggestion: Span,
-    pub start: String,
+    #[suggestion(code = "=", applicability = "machine-applicable")]
+    pub eq_suggestion: Span,
+    #[suggestion(code = "{literal}{suffix}", applicability = "machine-applicable")]
+    pub lit_suggestion: Span,
```
Comment (WaffleLapkin) [inline compiler/rustc_lint/src/lints.rs:1216]: This makes two suggestions, which indeed looks weird. Instead you should use multipart suggestions (you can search for `#[multipart_suggestion(` to find examples on how to do this).

Comment (MU001999) [reply compiler/rustc_lint/src/lints.rs:1216]: Got it! Thanks!

**Comment by scottmcm** [comment]
I'm not familiar enough with lint patterns to be a good reviewer here, so let's say

r? @WaffleLapkin 

since they commented already.

**Review by WaffleLapkin — APPROVED** [review]

**Comment by WaffleLapkin** [comment]
This looks fine to me, modulo `// run-rustfix` (I thought I added the comment ages ago but apparently no? anyway..,.)

File: tests/ui/lint/issue-109529.rs [hunk 4]
```diff
@@ -0,0 +1,4 @@
+fn main() {
```
[the remark this answers is not in the document]
Comment (WaffleLapkin) [reply tests/ui/lint/issue-109529.rs:3]: This is simply because sometimes we forget :')

**Review by WaffleLapkin — APPROVED** [review]

**Pull request merged**